pkg/trait/prometheus_test.go

```diff
@@ -0,0 +1,215 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trait
+
+import (
+	"context"
+	"testing"
+
+	"github.com/apache/camel-k/pkg/apis/camel/v1alpha1"
+	"github.com/apache/camel-k/pkg/util/kubernetes"
+	"github.com/apache/camel-k/pkg/util/test"
+
+	"github.com/stretchr/testify/assert"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestConfigurePrometheusTraitInRightPhaseDoesSucceed(t *testing.T) {
+	trait, environment := createNominalPrometheusTest()
+
+	configured, err := trait.Configure(environment)
+
+	assert.Nil(t, err)
+	assert.True(t, configured)
+}
+
+func TestConfigurePrometheusTraitInWrongPhaseDoesNotSucceed(t *testing.T) {
+	trait, environment := createNominalPrometheusTest()
+	environment.Integration.Status.Phase = v1alpha1.IntegrationPhaseResolvingKit
+
+	configured, err := trait.Configure(environment)
+
+	assert.Nil(t, err)
+	assert.False(t, configured)
+}
+
+func TestApplyNominalPrometheusTraitDoesSucceed(t *testing.T) {
+	trait, environment := createNominalPrometheusTest()
+
+	err := trait.Apply(environment)
+
+	assert.Nil(t, err)
+
+	container := environment.Resources.GetContainerByName(defaultContainerName)
+	assert.NotNil(t, container)
+	test.EnvVarHasValue(t, container.Env, "AB_PROMETHEUS_PORT", "9779")
+	ports := container.Ports
+	assert.Len(t, ports, 1)
+	assert.Equal(t, "prometheus", ports[0].Name)
+	assert.Equal(t, int32(9779), ports[0].ContainerPort)
+	assert.Equal(t, corev1.ProtocolTCP, ports[0].Protocol)
+
+	service := environment.Resources.GetService(func(service *corev1.Service) bool {
+		return service.Name == "integration-name"
+	})
+	assert.NotNil(t, service)
+	assert.Len(t, service.Spec.Ports, 1)
+
+	serviceMonitor := environment.Resources.GetServiceMonitor(func(service *monitoringv1.ServiceMonitor) bool {
+		return service.Name == "integration-name"
+	})
+	assert.NotNil(t, serviceMonitor)
+
+	assert.Len(t, environment.Integration.Status.Conditions, 1)
+	condition := environment.Integration.Status.Conditions[0]
+	assert.Equal(t, v1alpha1.IntegrationConditionPrometheusAvailable, condition.Type)
+	assert.Equal(t, corev1.ConditionTrue, condition.Status)
+}
+
+func TestApplyPrometheusTraitWithoutContainerDoesNotSucceed(t *testing.T) {
+	trait, environment := createNominalPrometheusTest()
+	environment.Resources = kubernetes.NewCollection()
+
+	err := trait.Apply(environment)
+
+	assert.Nil(t, err)
+
+	assert.Len(t, environment.Integration.Status.Conditions, 1)
+	condition := environment.Integration.Status.Conditions[0]
+	assert.Equal(t, v1alpha1.IntegrationConditionPrometheusAvailable, condition.Type)
+	assert.Equal(t, corev1.ConditionFalse, condition.Status)
+}
+
+func TestApplyPrometheusTraitWithServiceDoesNotSucceed(t *testing.T) {
```

Comment:
2 more eyes are welcome there. when applying the trait with an existing service, the status outcolme is 'IntegrationConditionServiceNotAvailableReason'. Correct ?

Reply:
I'm not sure about that.. maybe @jamesnetherton can shed some light..

Reply:
The `IntegrationConditionServiceNotAvailableReason` reason should only be set on the condition when the _service_ trait is disabled. Otherwise, this is a bug. Note that the current logic of the Prometheus trait may not be very robust as it makes the implicit assumption that only the _service_ trait can create a service beforehand.

Reply:
Just a small issue then, I've opened #1149 to keep track. I'll rebase, thanks for review :+1: